File: find_missing_monsters.js
/* eslint-disable @typescript-eslint/no-require-imports */
const fs = require('fs');
const https = require('https');

const LOCAL_FILE = 'src/lib/data/monsters_2024_final.json';
const OUTPUT_FILE = 'missing_monsters.json';

// Fetch Helper
function fetchUrl(url) {
    return new Promise((resolve, reject) => {
        const req = https.get(url, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => resolve(data));
        });
        req.on('error', reject);
        req.end();
    });
}

async function main() {
    try {
        // 1. Load Local
        let localSlugs = new Set();
        if (fs.existsSync(LOCAL_FILE)) {
            const localData = JSON.parse(fs.readFileSync(LOCAL_FILE, 'utf8'));
            localData.forEach(m => {
                if (m.slug) localSlugs.add(m.slug);
                const nameSlug = m.name.toLowerCase().replace(/[']/g, "").replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
                localSlugs.add(nameSlug);
            });
        }
        console.log(`Local monsters: ${localSlugs.size}`);

        // 2. Fetch Remote
        console.log("Fetching remote list...");
        const html = await fetchUrl("https://www.aidedd.org/dnd-filters/monsters.php");

        // Looser regex: look for the query param pattern irrespective of tag structure
        // Pattern: href="...monstres.php?vo=SLUG" or href='...'
        const linkRegex = /href=['"]([^'"]*monstres\.php\?vo=([^'"]+))['"]/gi;

        let match;
        const missing = [];
        let foundRemote = 0;

        // Also capture the name if possible. usually after the href closing tag >NAME</a>
        // But since we are doing a global scan on a massive string, we might just look for slugs first.
        // Let's iterate.

        // To get name, we really should match the <a> tag.
        // In the dump: <a href="...vo=slug" target='_blank'>Name</a>
        const tagRegex = /<a[^>]+href=['"]([^'"]*monstres\.php\?vo=([^'"]+))['"][^>]*>([\s\S]*?)<\/a>/gi;

        while ((match = tagRegex.exec(html)) !== null) {
            foundRemote++;
            let fullUrl = match[1];
            if (!fullUrl.startsWith('http')) {
                if (fullUrl.startsWith('../')) fullUrl = `https://www.aidedd.org${fullUrl.replace('..', '')}`;
                else if (fullUrl.startsWith('/')) fullUrl = `https://www.aidedd.org${fullUrl}`;
                else fullUrl = `https://www.aidedd.org/dnd/${fullUrl}`;
            }

            const slug = match[2];
            let name = match[3].replace(/<[^>]+>/g, '').trim(); // Remove inner tags if any

            if (!localSlugs.has(slug)) {
                missing.push({ name, slug, url: fullUrl });
            }
        }
        console.log(`Remote monsters found: ${foundRemote}`);
        console.log(`Missing monsters: ${missing.length}`);

        fs.writeFileSync(OUTPUT_FILE, JSON.stringify(missing, null, 2));
        console.log(`Saved to ${OUTPUT_FILE}`);

    } catch (e) {
        console.error(e);
    }
}

main();

File: find_missing_monsters.py
import requests
from bs4 import BeautifulSoup
import json
import re

# 1. Load Local Monsters
try:
    with open('src/lib/data/monsters_2024_final.json', 'r', encoding='utf-8') as f:
        local_data = json.load(f)
        # Slugs in local data might need normalization. 
        # The scraping process created slugs from names.
        # Let's see if we have 'slug' or 'url' in local data.
        # We constructed slugs as name.toLowerCase().replace(...) in the process script.
        # But let's look at the keys in 'monsters_2024.ts' or just re-generate slugs from names.
        local_slugs = set()
        for m in local_data:
            # Reconstruct the slug logic used previously
            slug = m['name'].lower().replace(' ', '-').replace("'", "")
            # Remove special chars
            slug = re.sub(r'[^a-z0-9-]', '', slug)
            local_slugs.add(slug)
            
            # Also add the URL-based slug if present
            if 'url' in m:
                url_slug = m['url'].split('/')[-1]
                local_slugs.add(url_slug)
except FileNotFoundError:
    print("Local file not found, assuming empty.")
    local_slugs = set()

print(f"Local monsters count: {len(local_slugs)}")

# 2. Fetch Remote List
url = "https://www.aidedd.org/en/rules/monsters/"
print(f"Fetching {url}...")
try:
    response = requests.get(url)
    if response.status_code != 200:
        print(f"Failed to fetch list: {response.status_code}")
        # Try the other URL found in browser logs if this fails
        url = "https://www.aidedd.org/dnd-filters/monsters.php" 
        print(f"Retrying with {url}...")
        response = requests.get(url)

    soup = BeautifulSoup(response.content, 'html.parser')
    
    # Selector based on browser investigation
    # 'a[href^="../dnd/monstres.php?vo="]'
    links = soup.select('a[href*="monstres.php?vo="]')
    
    remote_monsters = []
    for link in links:
        href = link.get('href')
        name = link.text.strip()
        # Extract vo parameter
        match = re.search(r'vo=([^&]+)', href)
        if match:
            slug = match.group(1)
            remote_monsters.append({'name': name, 'slug': slug, 'url': f"https://www.aidedd.org/dnd/monstres.php?vo={slug}"})

    print(f"Remote monsters found: {len(remote_monsters)}")

    # 3. Diff
    missing = []
    for rm in remote_monsters:
        if rm['slug'] not in local_slugs:
            # Try fuzzy match? (e.g. 'adult-black-dragon' vs 'dragon-black-adult')
            # For now, simplistic check.
            missing.append(rm)

    print(f"Missing monsters: {len(missing)}")
    
    # Save missing list
    with open('missing_monsters.json', 'w', encoding='utf-8') as f:
        json.dump(missing, f, indent=2)

except Exception as e:
    print(f"Error: {e}")
